### PR TITLE
Add GitHub Workflow to automate release generation

### PR DESCRIPTION
**Pull request by joanlopez**
~In order to align this _jslib_ with what was discussed (and consensuated) for [k6-jslib-summary](https://github.com/grafana/k6-jslib-summary/pull/8), and set up [here](https://github.com/grafana/k6-jslib-summary/pull/8), this pull request:~
~- Removes the `build` directory from control versioned files, in favor of automated releases~
~- Sets up a GH Workflow to automate releases from Git tags push.~
~- Updates the contribution guidelines accordingly.~

**UPDATE (08/01)**

It adds a new GitHub Workflow aiming the (future) automation of releases, heavily inspired by https://github.com/grafana/k6-jslib-summary/blob/main/.github/workflows/release.yml, but using Webpack & NPM instead of Yarn to adjust it to the project needs.

_I left some comments in form of open questions inline._

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 node_modules/
+build/
```
Comment (joanlopez) [inline .gitignore:2]: My only remaining question is: does the removal of the `build` folder makes new features no available until we tag & publish a new release? For instance, two months ago, [support for Lambda was added](https://github.com/grafana/k6-jslib-aws/pull/76), but no new release have been shipped since then.

So, in such case, I guess anyone could import directly from GitHub (does that work? is that correct?), but if we remove `build` directory, that will no longer be possible. Is that something we want? Or we're just fine with no one being able to use new features until they're shipped through a proper release?

Thanks!

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -51,21 +51,18 @@ In a PR:
 
 1. Bump the version in the `package.json` file.
 2. Run the `npm update` command to update the `package-lock.json` file.
-3. Run the `npm run webpack` command to ensure the build system produces the latest distributable files.
-4. Search and replace every occurences of the previous version in the `README.md` file with the new version.
-5. Search and replace every occurences of the previous version in the `/examples` directory with the new version.
+3. Search and replace every occurrence of the previous version in the `README.md` file with the new version.
+4. Search and replace every occurrence of the previous version in the `/examples` directory with the new version.
```
Comment (joanlopez) [inline CONTRIBUTING.md:55]: Now, code in `examples` directory doesn't work as-is, it requires running `npm run webpack` first.
Do you think that's fine? Or should we update the imports there to point to: `https://jslib.k6.io/aws`, and update the version in the import on each new release, as stated here?

File: .gitignore (hunk 1 above)
Comment (olegbespalov) [reply .gitignore:2]: :thinking: TBH, I assume that importing the library directly from GitHub won't work, and to get the library, you need to somehow _deliver_ it to your script. Either with `jslib.k6.io` or implement some process of cloning (or maybe importing with npm package if it's possible directly with github). 

On the other side, if a customer implements delivery of the library manually (sources download) and relies on the fact that the build is located in this directory, it could be a breaking change. Still, the case sounds hypothetical, but who knows :shrug: 

>For instance, two months ago, https://github.com/grafana/k6-jslib-aws/pull/76, but no new release have been shipped since then.

It's unavailable for most users via `jslib.k6.io`, but if the customer sets up the process of importing code directly from GitHub, it's available.

TL;DR: since the project already has a `build,` and you're right that somebody could rely on the existence of the `build` directory, it's unsafe just deleting it. But the `release` workflow is still essential and could help us with the automation of distribution builds to the `jslib.k6.io`.

Comment (oleiade) [reply .gitignore:2]: > My only remaining question is: does the removal of the build folder makes new features no available until we tag & publish a new release?

For context, this is the reason I kept the `build` folder so far, as it allowed to point users to build files in between releases when a fix was not yet released for instance. My understanding is that one could pick the address of the raw build file from GitHub, import it in their script, and it would work (haven't tried it myself, but I assume that works).

From the development of the library's perspective, I think it was convenient to have that folder around. On the other hand, the gain it brought (not having to download the whole library, and build the files for a given commit yourself while developing), is also counterweight-ed by the noise it adds to PRs by including the build files there (which can be mitigated by a CI step as we discussed, but still :)).

I have no strong opinions either way on this one, thus feel free pick to pick the way you judge best 🤝

File: CONTRIBUTING.md (hunk 2 above)
Comment (oleiade) [reply CONTRIBUTING.md:55]: I think we should instead mention the `npm run webpack` step, as I would prefer examples to work "locally" for each commit.

**Comment by oleiade** [comment]
As we're at it, I thought of renaming the `webpack` command to `build`, to align it with what I've observed in the ecosystem. It's a rather small change I believe, but it would also help "hide" that we rely on webpack which feels like an implementation detail.

What do you think?

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+      - name: Check out source code
+        uses: actions/checkout@v4
+      - name: Use Node.js
+        uses: actions/setup-node@v4
+        with:
+          node-version: '18.x'
+      - name: Install dependencies
+        run: npm install
+      - name: Build & archive
+        run: npm run webpack && tar -zcf build.tar.gz build
```
Comment (oleiade) [inline .github/workflows/release.yml:22]: I would advocate for an explicit name here: `k6-jslib-aws-0.11.0.tar.gz`. So I know what that archive contains, just by looking at its name when I find it back six months from now in my downloads folder 😄 

What do you think?

Comment (olegbespalov) [reply .github/workflows/release.yml:22]: I see your point about the naming to help discovery in local folders. However, I still prefer to keep the name since it will be the convention between the sub-repositories, and the process of looking for the release build for the future porting to the jslib.k6.io is much simpler and less error-prone.

Comment (oleiade) [reply .github/workflows/release.yml:22]: Fair enough 👍🏻 Let's go for that then.

Comment (olegbespalov) [reply .github/workflows/release.yml:22]: >humans using the releases in the process too

I can't add more. That's precisely what I'm trying to do. In the end, the end goal of this process is to align all sub-repositories and improve the delivery of the builds to the `jslib.k6.io` since it's the most straightforward place (and probably) the most convenient for humans.

Comment (oleiade) [reply .github/workflows/release.yml:22]: I didn't mean to argue @olegbespalov, I agree the effort to improve `jslib.k6.io` will be beneficial (to humans too ^^) 👍🏻

Comment (joanlopez) [reply .github/workflows/release.yml:22]: > [...] the process of looking for the release build for the future porting to the jslib.k6.io is much simpler and less error-prone.

I'm happy to keep it as-is, of course, but could you clarify this, please? @olegbespalov 

<br/>
Honestly, I mostly copied the workflow definition from k6-jslib-summary, but now that @oleiade has suggested it, I feel like it could be a fair suggestion to consider. My main thoughts around that are:

- If there's any manual step in the process of publishing it (like creating the PR in [jslib.k6.io](https://github.com/grafana/jslib.k6.io) repo), isn't it less error prone to explicitly see the version you're manipulating the files of?

- If we care about the directory name (I see in [jslib.k6.io](https://github.com/grafana/jslib.k6.io) repo it's just the version number), couldn't we generate a compressed file name like `k6-jslib-aws-0.11.0.tar.gz`, that contains the `0.11.0` folder with all the files?

Thanks! 🙇🏻

**Comment by joanlopez** [comment]
> As we're at it, I thought of renaming the `webpack` command to `build`, to align it with what I've observed in the ecosystem. It's a rather small change I believe, but it would also help "hide" that we rely on webpack which feels like an implementation detail.
> 
> What do you think?

I prepared that in the first pull request I prepared around this, but then I did care less because in this approach the folder isn't being commited. But yeah, it's a minor but fair change, that I can easily add in this pull request. Thanks!

File: CONTRIBUTING.md (hunk 2 above)
Comment (joanlopez) [reply CONTRIBUTING.md:55]: Okay, let me try to slightly modify docs to make such thing clear 👌🏻

File: .github/workflows/release.yml (hunk 3 above)
Comment (olegbespalov) [reply .github/workflows/release.yml:22]: @oleiade Just in case, I haven't taken this as arguing and didn't argue. I just wanted to provide the complete picture of the idea behind the name.

Comment (oleiade) [reply .github/workflows/release.yml:22]: Sounds good I appreciate it 🙇🏻 It's always difficult for me to decipher intentions, tone, and "read the room" in written online discussions, so I prefer to be safe than sorry 🤝

Comment (olegbespalov) [reply .github/workflows/release.yml:22]: @joanlopez for both your thoughts it's true for the current state of the things. 

But the main idea is to get rid of this manual steps. Or at least for beginning script them and end it up with a simple command (in `jslib.k6.io`) that could do most of the routine which are:

1) downloads an asset file of a certain release from the release `https://github.com/grafana/{pkgRepo}/releases/download/{pkgVersion}/build.tar.gz`
2) extracts this under the `lib/{pkgName}/{pkgVersion}`
3) updates `supported.json` and index.html
n) another routine that could be automated

**Review by olegbespalov — COMMENTED** [review]
TBH, after a discussion in treads, I think that maybe removing the `build` directory for `k6-jslib-aws` isn't a good thing (or perhaps it is too early to decide). 

Also, I feel that the idea of porting the changes to the `jslib.k6.io` isn't completely clear without a prototype (a PR to the `jslib.k6.io`) that could demonstrate the new flow.

How about putting this PR on pause? I'll sketch up the PR with the routine automation I mentioned in https://github.com/grafana/k6-jslib-aws/pull/89#discussion_r1441564275, and we can continue the discussion.

File: .github/workflows/release.yml (hunk 3 above)
Comment (olegbespalov) [reply .github/workflows/release.yml:22]: Actually, writing things down helps as usual, and I see no issues downloading files from `https://github.com/grafana/{pkgRepo}/releases/download/{pkgVersion}/{pkgRepo}-{pkgVersion}.tar.gz`, and the file generates automatically so if you folks feel that it's improvement let's go this way :+1:

**Comment by joanlopez** [comment]
Hey @olegbespalov, @oleiade,

After our discussion, and as per what's commented in https://github.com/grafana/k6-jslib-aws/pull/89#discussion_r1441580456, I decided to update this pull request to just add the release workflow, so at least we can turn that step into a 1-click automation (it just requires creating the tag) and move on.

Wdyt?

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        uses: actions/setup-node@v4
+        with:
+          node-version: '18.x'
+      - name: Install dependencies
+        run: npm install
+      - name: Build & archive
+        run: |
+          set -x
+          VERSION="${GITHUB_REF##*/}"
+          npm run webpack && tar -zcf "aws-${VERSION}.tar.gz" -C build .
```
Comment (joanlopez) [inline .github/workflows/release.yml:25]: It only includes (`-C`) the file within the directory to the compressed file (but not the directory itself), which I guess is fine, and may make our future simpler because:
- It prevents "breaking changes" based on the directory name (in the future we can switch `build` with `dist` easier).
- It makes the job of future automations more flexible, cause we won't need to manipulate directories (like renaming it or similar), we'll just need to create the folder where we want to place the files and uncompress (we can always use the `-C` flag again, if needed).

And, I don't expect in the foreseeable future any reason why having a recursive tree within the directory, which is probably the only "downside" of this approach?

Comment (olegbespalov) [reply .github/workflows/release.yml:25]: The `-C` flags what is missing even in the `k6-jslib-summary`; I learned that hardly when I was trying to sketch up a prototype of the semi-automation for the `jslib.k6.io`, so :+1:

Comment (olegbespalov) [reply .github/workflows/release.yml:25]: But I'd like to ask to change the name to the `repo-id`-`version,` e.g., `k6-jslib-aws-v0.0.0.tar.gz`. 

Just the `aws` could mislead that inside something from `aws` or too generic. And using the just repository ID helps us to avoid introducing and maintaining an additional mapping where we need to figure out that the asset of the `k6-js lib-aws` has a name `aws.`

What do you think?

Comment (joanlopez) [reply .github/workflows/release.yml:25]: > But I'd like to ask to change the name to the `repo-id`-`version,` e.g., `k6-jslib-aws-v0.0.0.tar.gz`.

That's absolutely fine for me! Indeed, I had the doubt about what naming convention to use, and I decided to use just `aws` because that's what's reflected in the [https://github.com/grafana/jslib.k6.io](https://github.com/grafana/jslib.k6.io), but I also agree that generally speaking is probably more clear, less error prone, to use the full name.

Any thoughts other thoughts? @oleiade

Comment (oleiade) [reply .github/workflows/release.yml:25]: Hey folks, I'm perfectly happy with those changes 👍🏻 🙇🏻

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+      - name: Create release with assets
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          set -x
+          VERSION="${GITHUB_REF##*/}"
+          gh release create "${VERSION}" "aws-${VERSION}.tar.gz" "build/*" --target "${GITHUB_SHA}" --title "${VERSION}"
```
[the remark this answers is not in the document]
Comment (oleiade) [reply .github/workflows/release.yml:32]: Happy with it, go for it 👍🏻

**Review by oleiade — APPROVED** [review]

File: .github/workflows/release.yml (hunk 4 above)
Comment (joanlopez) [reply .github/workflows/release.yml]: Thanks! Done! cc/ @olegbespalov

**Review by olegbespalov — APPROVED** [review]
